CRM/Core/Resources.php

```diff
@@ -272,21 +283,28 @@ public function addVars($nameSpace, $vars) {
    * Extensions and components should generally use addVars instead.
    *
    * @param array $settings
+   * @param string $region
+   *   The region to add settings to (eg. for payment processors usually billing-block)
+   *
    * @return CRM_Core_Resources
    */
-  public function addSetting($settings) {
-    $this->settings = $this->mergeSettings($this->settings, $settings);
-    if (!$this->addedSettings) {
+  public function addSetting($settings, $region = NULL) {
+    if (!$region) {
       $region = self::isAjaxMode() ? 'ajax-snippet' : 'html-header';
-      $resources = $this;
-      CRM_Core_Region::instance($region)->add([
-        'callback' => function (&$snippet, &$html) use ($resources) {
-          $html .= "\n" . $resources->renderSetting();
-        },
-        'weight' => -100000,
-      ]);
-      $this->addedSettings = TRUE;
     }
+    $this->settings = $this->mergeSettings($this->settings, $settings);
+    if (isset($this->addedSettings[$region])) {
+      return $this;
+    }
```

Comment:
I'm confused about this bit. Is `$this->settings` global to all regions? So adding a setting to one region actually adds it to all regions that get rendered? Should we instead work with `$this->settings[$region]`?

Reply:
It's especially confusing due to the early return that's not so early that line 295 doesn't happen first. Why is that?

Reply:
> It's especially confusing due to the early return that's not so early that line 295 doesn't happen first. Why is that?

That confused me too! We are "rendering" a callback so the actual callback does not run until all settings have been added (ie. multiple calls to this function, multiple calls to `$this->mergeSettings` on line 295). But we don't want to add the callback multiple times so we follow with an early return if it's already added.

`$this->settings` is global. We don't actually care what region they are added too as long as they are added. The only time we could end up with duplicates is if `html-header` was rendered after something else - but then the javascript breaks anyway because `CRM is not defined`. Otherwise we are always extending the CRM object so any already-rendered settings would be overwritten (which is what we want). For example a payment processor on initial load adds it's settings to the form, the payment processor is changed and they are overridden by the new settings loaded in through the billingblock.

Reply:
Got it!